src/Features/Core/Portable/EmbeddedLanguages/AbstractLanguageDetector.cs

```diff
@@ -171,16 +173,48 @@ private bool IsEmbeddedLanguageStringLiteralToken(SyntaxToken token, SemanticMod
         }
     }
 
-    internal abstract class AbstractLanguageDetector<TOptions, TTree> :
+    internal interface ILanguageDetectorInfo<TDetector>
+    {
+        ImmutableArray<string> LanguageIdentifiers { get; }
+        TDetector Create(Compilation compilation, EmbeddedLanguageInfo info);
+    }
+
+    internal abstract class AbstractLanguageDetector<TOptions, TTree, TDetector, TDetectorInfo> :
         AbstractLanguageDetector<TOptions>
         where TOptions : struct, Enum
         where TTree : class
+        where TDetector : AbstractLanguageDetector<TOptions, TTree, TDetector, TDetectorInfo>
+        where TDetectorInfo : struct, ILanguageDetectorInfo<TDetector>
```

Comment:
by passing the info as that interface *and* as a struct, the base type can just do `default(TDetectorInfo).Whatever` and have that work.